### DISCUSSION TITLE
[6기] 모바일 안드로이드 레벨4 바퀴의 재발명 | 2024.09.10

### DISCUSSION THREAD

**Discussion by woowahan-leah** [woowacourse/retrospective#54] [General]
"[바퀴의 재발명(To reinvent the wheel)](https://en.wikipedia.org/wiki/Reinventing_the_wheel)"이라는 단어를 들어본 적이 있나요? 이 관용구는 **이미 발명되었고 본질적인 결함이 없는 것을 다시 발명하는 것은 가치를 더하지 못하고 시간 낭비**일 뿐이라는 의미로 "**바퀴를 재발명하지 말라**"라고 말하는 데 주로 사용됩니다.
하지만 **때로는 "바퀴를 재발명"하는 것이 필요**할 때도 있습니다. 어떤 상황에서 바퀴를 재발명하는 행위가 필요할까요? **재발명을 할 때 얻을 수 있는 이점과 단점에 대해 생각해 보고 1,000자 이내로 작성**해 주세요. 수업 시간에 각자가 쓴 내용을 함께 토론해 볼게요.

**Comment by haeum808** [comment]
재발명의 이점
- 내 자동차에 더 적합한 바퀴를 만들면서 불필요한 부분은 떼어낼 수 있다.
- 바퀴의 비용을 올린다면 직접 개발하여 해결할 수 있다.
- 바퀴를 만들고 나면 다른 종류의 바퀴도 쉽게 만들 수 있다.

재발명의 단점
- 시간과 비용이 많이 소모된다.
- 바퀴를 새로 만들었는데 기존 바퀴와 비슷하거나 수준이 낮을 수 있다. 혹은 아예 만들어내지 못할 수 있다.

**Comment by Hogu59** [comment]
이점)
1. 현재 진행하는 프로젝트, 혹은 팀의 서비스에 가장 적합한 형태로 서비스를 구현할 수 있다.
2. 서비스를 운영하는 도중, 필요한 부분을 따로 수정할 수 있다. (수정에 대한 유연성 확보)
3. (라이브러리와 같이 경우라면; 뇌피셜) 이후 버전에 대한 지원 중단 혹은 지원 종료 등에 대해 염려하지 않아도 된다.

단점)
1. 개발 리소스가 들어간다. (개발 내용에 따라 많이 들어갈 가능성이 높다.)
2. 기존의 라이브러리, 혹은 기능보다 많은 버그 발생 가능성을 갖는다.
(전제조건; 비교 대상인 라이브러리가 충분히 많은 사용자를 통해 버그를 수정했다는 전제)

**Comment by aprilgom** [comment]
바퀴도 개념은 단순해보이지만, 지금의 바퀴를 만들기 까지 수 많은 바퀴가 만들어졌을거라 생각합니다. 바퀴의 발명은 기원 전 3500년이었지만, 타이어가 발명된 건 1847년이었죠. 똑같은 걸 만드는 건 분명 낭비처럼 보이지만, 재발명을 통해 개선점을 발견할 기회가 생기는 것 아닐까요.
그리고, 어떤 불편함에서 어떤 개선점들이 나왔는지 맥락을 알고 원리를 알면 이후에 일반적이지 않은 문제.. 남이 만들어놓은 바퀴로는 해결이 안될 때 남이 만든 바퀴를 조금 고치던 아예 직접 만들던 해결할 가능성이 높을 것 같습니다. 재발명을 하지 않는다면 문제의 접근방법도 모르는 상태로 있지 않을까요?

**Comment by kimhm0728** [comment]
저는 기존 바퀴가 특정 상황(언어, 도메인 등)에서만 적합한 경우나 성능/보안 상 문제가 있는 경우 바퀴를 재발명하려고 할 것 같습니다.
바퀴를 재발명한다면 바퀴에 대한 이해도를 더 높일 수 있습니다. 다시 기존의 안정적인 바퀴를 사용하더라도 더 잘 활용하고, 효율적으로 사용할 수 있습니다.
단점으로는 개발 리소스가 너무 많이 들며, 버그가 발생할 수 있습니다. 이미 안정적으로 사용되고 있는 기술보다는 불안정적으로 작동될 가능성이 훨씬 높습니다.

**Comment by gaeun5744** [comment]
저희는 맥북의 동작 원리를 몰라도 코딩을 할 수 있고, 자동차의 원리를 몰라도 운전을 할 수 있습니다.
이미 발명된 물건들을 이용해서 쉽고 빠르게 편안한 삶을 누리고 있죠. 

하지만 기존의 발명품도 완벽할 수는 없습니다. 우리 스스로도 알아차리지 못하는 불편함이 있죠.
저는 이를 알아가는 과정이 `재발명`이라고 생각합니다. 어떠한 기능이 왜 생겨났는지 알아가고, 본인의 경험에 비추어 짚어가는 과정이 개선의 첫걸음이라 생각합니다.

무엇이든 기존의 문제를 제대로 파악해야 더 나은 결과가 나오고, 기존의 기술을 제대로 활용할 수 있습니다.

하지만 단점으로는, 과연 그 `재발명`의 과정이 성공적인가?에 대한 불확실성이라 생각합니다.
기존의 익숙한 사고에 갖혀서 동작원리를 이해하는데에 그칠 수 있고, 새로운 문제를 발견할 수 있습니다.

시간 투자를 통해 얻을 수 있는 가치와 기존을 방식의 따르는 것에 대한 시간 절약을 비교해야하기 때문에, 언제 재발명이 필요한가?는 너무 어려운 문제 같습니다.

**Comment by ii2001** [comment]
바퀴를 재발명 하는 행위가 필요할 때
- 바퀴가 불편해질때 or 바퀴가 굴러가는게 지루해 졌을때

재발명을 해서 얻을 수 있는 장단점
- 장점
    - 재밌음
    - 혁명적인 방법을 생각해낼 가능성 있음
- 단점
    - 매우 귀찮음
    - 쓸모없는 시간을 많이 씀
    - 주변에서 뭐하냐고 핀잔을 매우 많이 받을 수 있음

**Comment by chaehyuns** [comment]
이번 과제를 통해 가장 크게 느낀 장점은 ‘기술 이해도의 증진’입니다. 
그동안 개발자로서 라이브러리나 툴을 빠르게 사용하는 데에만 집중했다면, 이번에는 기술이 어떻게 구현되어 있는지 깊이 고민해 볼 기회를 가질 수 있었습니다. 
기존의 기술을 직접 구현해보면서 로직에 대해 이해하는 데 큰 도움이 되었습니다. 
단순히 기존의 기술을 사용하는 것에 그치지 않고 스스로 개선하고 더 나은 방향으로 발전시킬 수 있는 방법을 찾을 수도 있겠다는 생각을 넓히는 경험이 되었습니다. 
단점은 이미 구현되어 있는 기존의 것을 사용하지 않는다면, 진짜 필요한 기능을 구현하는 데에 시간을 활용할 수 없다는 점입니다. 
그러나 앞으로도 기술을 단순히 사용하는 데 그치지 않고 그 이면의 원리와 설계를 깊이 이해해보는 경험도 필요했다는 생각이 들었습니다.

**Comment by JoYehyun99** [comment]
재발명의 목적에 따라 그 가치가 달라질 것이라고 생각합니다.
학습이나 개선을 목적으로 한다면 충분한 이점이 있을 것 같습니다. 재발명 과정을 통해 원리를 파악하고 깊이 이해할 수 있으며, 그 과정에서 더 발전된 결과물이 나올 가능성도 있습니다.
하지만 이러한 목적이 없다면, 이미 효율적인 대안이 존재하는 상황에서의 재발명은 시간과 비용을 낭비할 수 있다는 단점이 더욱 클 것 같습니다.

**Comment by murjune** [comment]
# 바퀴를 재발명한 경험
`이미지 처리`의 심화적인 작업이 요구될 때마다 막막함을 느꼈다.
Glide 내부 코드를 봐도 도저히 이해가 가지 않았기에, 직접 바퀴를 만들었다.

바퀴를 만들며 학습하니 막혔던 문제들이 차츰 해결되었고, Glide 가 어떻게 동작하고, 왜 그렇게 설계되었는지도 알 수 있었다.

내가 만든 바퀴는 `찌그러진 바퀴`였지만 `이미지 처리` 작업은 이제 `두려움`에서 `자신감`의 대상이 되었다.
이후 어떤 바퀴를 사용하든지 빠르게 적응하여 수월하게 작업을 할 수 있게 되었다.

반면, ＇바퀴의 재발명`은 시간과 노력을 많이 요구하는 작업이라는 것도 알았다.
바퀴에 대한 배경지식이 하나도 없으면 학습해야 할 것이 끝없이 많다.

>  사실은 URL로 이미지를 받아오는 거였구나.
> 그런데 네트워크 통신은 어떻게 하지?
> InputStream? Bitmap?  LRU 캐시?

바퀴의 재발명은 의미 있는 경험이지만, 동시에 매우 힘들고 고단한 작업이다.
이 작업이 나에게 충분한 가치가 있는지 고민해보자.
만약, 바퀴를 재발명하겠다는 결심을 했다면 모든 것을 다 알아야겠다는 생각도 멈추길 바란다.  
어차피 다 알 수 없고, 개발이 싫어질 수도 있다.

**Reply by murjune** [reply]
회고 느낌으로 써봤습니당 ㅎ ㅎ

**Comment by jinuemong** [comment]
이번 미션을 수행하고난 후 `라이브러리를 제대로 이해하고 사용하고 있는가?` 에 대한 생각을 하게 되었습니다. 그동안 개발자로서 라이브러리를 사용할 때, 어떤 기능을 하는지에 초점을 두고 개발을 진행하였습니다. 빠른 개발이 가능하였지만, 실제로 `어떻게 동작하는지` 에 대한 질문을 받았을 때 답변하기 어려웠습니다. 사용한 라이브러리가 무슨 기능을 하는지 설명할 수는 있지만, 어떻게 동작하는지에 대해 설명할 수 없다는 것을 알게 되었습니다. 

미션을 통해서 기술을 직접 구현해보면서 로직을 이해하였고, 어느 정도는 설명이 가능해졌습니다.

바퀴를 재발명 했을 때 해당 구성을 잘 이해할 수 있고, 불필요한 부분을 제거하여 사용할 수 있음을 알게 되었습니다. 해당 바퀴로 인해 파생되는 다른 바퀴의 제작도 가능해졌고, 기능을 잘 이해하는데 도움이 되었습니다.

반면에 시간과 비용이 많이 소모되었습니다. 내가 잘 이해하고 설명하는 것이 가능해졌지만 `내 바퀴과 기존의 바퀴를 대체할 수 있을까?` 에 대한 생각을 하게 되었습니다.

**Comment by kmkim2689** [comment]
바퀴를 재발명한다는 것은 특정 사고를 하는 사람에게 대박으로 다가올 수 있는 `가챠`라고 생각합니다.

다수의 사람들이 어떤 것을 `재발명한다`라는 행위를 할 때, 모두가 완전히 같은 방식으로 동일한 것을 만들지는 않을 것입니다.
그대로 설계서를 보고 따라 만들지 않는 이상, 비슷한 결과물이어도 각 결과물에는 각자의 생각이 녹아들어 있을 것입니다.
멀리 갈 것도 없이, 이미 소스코드가 존재하는 프로그램을 다시 만들어내야 하는 미션에서 크루들은 각자 다른 코드를 작성했습니다.

그 과정과 결과 속에서, 이전에 없었던 것이 생겨날 수도 있습니다.
더 나아가, 이전에는 맞았던 것이 틀렸다는 것으로 판명날 수도 있습니다.
무조건 맞다고 생각했던 사상 및 가설이 수도 없이 뒤집어지는 과정 속에서, 인류는 끊임없이 발전해오고 있습니다.

특정 분야를 깊이 연구하는 것이 아닌 이상, 
불필요한 부분까지 고려하여 어떤 것을 다시 만든다는 것은 누군가에게는 시간낭비일 수 있습니다.
또한 미지의 분야를 발명을 통해 알아간다는 것은 누군가에게는 고난의 시간이 될 것입니다.

다만 바퀴를 재발명한다는 것은 `혹시나`하는 의심으로부터 시작했지만,
특정 분야의 역사와 판도를 바꿔버릴 수도 있는 로또와도 같은 기회라고 볼 수 있다고 생각합니다.
이러한 가챠의 연속이 우리가 몸담고 있는 이 분야의 획기적인 발전을 이끌어 낼 것이라고 생각합니다.

**Comment by songpink** [comment]
"바퀴를 재발명하지 말라" 라는 프로그래밍 격언은, 이미 잘 만들어진 라이브러리가 있을 경우 굳이 새로운 라이브러리를 만들지 말라는 말일 것입니다. 이미 만들어진 라이브러리를 사용하는 것이 시간과 노력을 절약할 수 있고, 많은 동료 개발자들과도 소통하기 쉬우며, 더 좋은 프로그램을 만들 수 있기 때문입니다.

하지만 그럼에도 불구하고 현재 우테코 레벨4에서는 바퀴를 재발명하는 미션을 하도록 하고 있습니다. 바퀴를 재발명 하는 행위가 전혀 무의미한 것은 아니기 때문입니다. 바퀴를 다시 만들어 보는 과정에서 바퀴의 원리를 이해할 수 있고, 원리를 이해하였기 때문에 더 좋은 방식으로 바퀴를 사용할 수 있습니다. 바퀴를 직접 만들어 본 사람은 그렇지 않은 사람보다 바퀴가 어느정도의 하중을 견딜 수 있고, 어떻게 달아야 가장 잘 굴러가는지를 알고 있습니다. 이 둘은 바퀴를 사용하는 방법에 있어서 차이가 발생할 것입니다. 따라서 학습에 있어서 바퀴를 재발명 하는 행위는 의미가 있다고 생각합니다.

하지만 그것을 우리의 프로젝트에 사용하기 위해 재발명 하는 것이라면, 이미 만들어 져 있는 바퀴를 사용하는 것이 더 효율적이고 효과적인 방법일 것입니다.

**Comment by hxeyexn** [comment]
### 재발명하는 행위가 필요한 상황
기존 바퀴의 틀에서 벗어나 자유로운 변형이 요구되거나 기존 바퀴 원리에 대한 분석이 필요할 때, 재발명하는 행위가 필요하다고 생각합니다. 

---

### 이점
**1. 동작 원리 이해**
- 재발명에 실패하더라도 이를 통해 동작 원리를 이해하게 된다면 큰 이점을 얻을 수 있다고 생각합니다. 
재발명을 위해서는 기존 기술의 동작 원리를 깊이 이해해야 합니다. 
이 과정에서 그 원리를 파악하게 되면 해당 기술의 이점을 더 효과적으로 활용할 수 있습니다.

**2. 최적화**
- 자동차를 직접 개발한다면, 기존의 바퀴를 사용하는 것보다 바퀴를 재발명했을 때 자동차의 성능을 최적화할 수 있습니다.

**3. 자유로운 변형**
- 직접 발명한 바퀴는 비교적 자유롭게 변형할 수 있습니다. 이를 통해 다양한 종류의 자동차를 보다 수월하게 개발할 수 있습니다.    

---

### 단점
**1. 창의적 한계**
- 기존에 정의된 바퀴의 틀에 갇혀 더 나은 아이디어가 떠오르지 않을 수도 있습니다.
이는 기존의 바퀴를 단순히 모방하는 데 그쳐, 더 우수한 바퀴를 발명하지 못하는 결과로 이어질 수 있다고 생각합니다.

**2. 시간, 비용 부담**
- 재발명에는 많은 비용과 시간이 소요됩니다. 자동차 개발 과정에서 바퀴를 재발명하는 데 대부분의 시간을 할애하다보면, 
최종 목표인 자동차는 퀄리티가 떨어지거나 정해진 시간 내에 완성하지 못할 수 있습니다.  

**3. 불확실한 안정성**
- 기존의 바퀴는 오랜 시간 동안 검증되어 왔습니다. 
그러나 재발명된 바퀴는 시간이 지남에 따라 어떤 문제점이 발생할 지 예측하기 어려워 불안정할 수 있습니다.

**Comment by dpcks0509** [comment]
개발 초기에는 코드가 잘 돌아가기만 하면 되었고, 코드 구조나 동작 원리에는 큰 관심이 없었습니다. 그러나 문제가 발생했을 때, 내부 원리를 알지 못해 해결하는 데 시간이 오래 걸리곤 했습니다.

바퀴를 재발명하는 방식으로 직접 동작 원리를 파악하면 시스템을 더 잘 이해하고, 문제를 빠르게 해결할 수 있습니다. 또한, 이를 바탕으로 새로운 아이디어를 떠올릴 수 있는 장점도 있습니다.

하지만 바퀴를 재발명하는 데는 불필요한 시간과 자원이 소모될 수 있으며, 이미 검증된 방법을 놔두고 새로운 해결책을 찾는 것은 비효율적일 수 있습니다.

따라서 바퀴를 재발명하는 것은 상황에 따라 필요할 수 있지만, 그 필요성과 비용을 신중하게 고려해야 한다고 생각합니다.

**Comment by kkosang** [comment]
사용자의 요구사항이 변경되어 기존 시스템(바퀴)도 변경되어야 할 때 재발명하는 것이 필요합니다.
재발명을 통해 변경된 요구사항을 만족시키거나 시스템 결함을 해결할 수 있습니다.
더 나아가 기존의 시스템보다 좋은 품질을 만들 수 있습니다.
현재 우리가 사용하고 있는 시스템은 한 번에 나온 것이 아닌, 누군가가 꾸준히 재발명하고 개선했기 때문에 나온 것이라고 생각합니다.

하지만 바퀴를 재발명하는 것에는 단점도 존재합니다.
새로운 요구사항에 맞춰 시스템을 다시 설계하고 발명하는데 많은 시간과 비용이 소모됩니다.
많은 시간과 비용을 소모했음에도 기존의 시스템보다 품질 저하가 발생할 수도 있습니다.

따라서 재발명 하려는 목적과 리스크를 고려하여 결정해야 합니다.

**Comment by Junyoung-WON** [comment]
이번 `DI 라이브러리를 직접 구현하기` 미션을 하며 느꼈고, 또 혼자 곰곰히 생각해본, **`바퀴의 재발명`** 의 이점과 단점입니다!

## 이점
### 학습의 측면에서 큰 도움이 됩니다!
과거의 어떤 기술자들이 "본질적 결함"이 없는 발명을 만들어갔던 과정을 처음부터 몸소 배우고 느끼며, 그 발명품에 대해서 자세하게 알 수 있게 됩니다.
왜 발명을 하게 되었는지와, 발명품이 왜 지금의 형태를 지니고 있는지에 대한 본질적인 이유를 깨달을 수 있습니다.
이렇게 재발명하는 과정에서 필요한 추가적인 지식과 기술을 터득하게 됩니다. 기술에 대한 이해도와 개발 능력을 크게 향상시킬 수 있는 좋은 학습 방법이라고 생각합니다.

### 우리의 상황과 환경에 적절한 새로운 바퀴를 만들 수 있습니다.
만약 평지에서 사용하던 바퀴를 가져다가 길이 험한 비탈길 또는 산길에서 사용하게 된다면 어떤 문제가 발생할 수 있을까요?
울퉁불퉁한 길바닥 때문에 잘 굴러가지 않거나, 쉽게 고장나거나 부서질 위험이 있습니다.
이런 환경 속에서 바퀴를 재발명 한다면, 기존의 바퀴보다 더욱 견고하고 튼튼하며, 험한 길에도 잘 굴러갈 수 있는 새로운 형태의 바퀴를 발명하게 될 수도 있습니다.
즉, 특정 환경이나 상황에서 더 잘 동작하는 자신만의 특별한 도구를 만들어낼 수 있을 것입니다.
그렇게 재발명한 발명품을 우리끼리 잘 활용한다면, 기존의 바퀴 보다도 더 가치있는 새로운 발명품이 되지 않을까요?

## 단점
### 우리의 시간과 자원은 소중합니다.
발명 자체에는 적지않은 시간과 노력이 필요합니다. 하지만 우리가 가진 자원은 한정적이에요.
잘 만들어진 발명품을 사용하지 않고 굳이 처음부터 재발명하게 된다면, 다른 것을 만들어나갈 자원을 낭비하게 될 수 있습니다.
자동차를 만들어달라고 했는데, 가장 중요한 엔진을 제쳐두고 바퀴부터 만들게 된다면, 엔진에 투자할 수 있는 자원과 시간을 뺏기게 됩니다.
한정된 자원과 시간 속에서 완성된 자동차는 그 엔진의 성능이 고객의 기대에 못 미칠지도 모릅니다.

### 발명 자체가 어렵습니다.
"결함" 없는 발명품을 만들기란 정말, 매우, 아주 어려울 것 같습니다.
오래 전부터 꾸준히 사용되어 왔고, 다른 더 좋은 대안책이 나타나고 있지 않는다는 것은 그만큼 잘 만든 발명품이란 증거입니다.
수없이 많은 시도와 시간, 노력이 녹아들어간 훌륭한 발명품을 처음부터 스스로 만들기란 쉽지 않을 것입니다.
구현이 아닌 학습의 목적이라 하더라도, 학습 자체의 난이도가 너무 높을 수 있습니다.
상황에 따라 더 적절하거나 효율적인 다른 학습 방법이 충분히 많습니다. 무작정 바퀴를 재발명하는 것은 오히려 학습을 방해하는 요인이 될 수 있습니다.

### 이상한 바퀴를 만들 수도, 이를 모르고 사용할 수도 있습니다.
바퀴를 발명하게 된 문제 상황을 자세히 이해하지 않거나, 바퀴에 관련된 지식과 기술들을 잘 모른다면, 제대로 굴러가지 않는 요상한 바퀴를 만들고 이를 사용할 수 있습니다.
바퀴를 발명한 이유는, 물건 따위를 싣는 물체에 장착하여 원활하게 물건을 운송하기 위함 입니다.
바퀴를 발명하게 된 이유를 충분히 이해하지 않고 무작정 바퀴를 발명한다면, 충족해야 할 요구 사항을 제대로 만족하지 못하고 잘 굴러가지 않는 어설픈 바퀴가 될 수 있습니다.
게다가 바퀴에 대한 지식과 발명에 필요한 기술이 부족하다면, 추가적인 개선 방안이 있음을 인지하지 못할 수 있습니다.
바퀴는 여러 환경에서도 쉽게 운송할 수 있도록 잘 굴러가야 하기에, 마찰력, 내구성 등 다양한 점을 고려하여야 합니다. 이에 대한 지식이 없다면 얼마 가지 못해 고장나거나 망가질 바퀴가 만들어질 수 있습니다.
심지어 그 지식이 너무 부족하다면 바퀴의 존재 자체를 모를 수도 있습니다. 바퀴라는 훌륭한 해결책의 존재를 모르고서, 어설프게 바퀴를 발명하고 이를 사용하게 되어 후에 큰 문제점을 야기할 수 있습니다.
결국 기존의 바퀴를 사용하느니만 못하게 됩니다!

**Comment by junjange** [comment]
재발명을 했을 때의 이점으로는, 발명가의 생각과 의도를 보다 깊이 이해할 수 있다는 점이 있을 것 같아요. 비록 온전히 파악하지는 못하더라도, 재발명 과정을 통해 그 발명품의 동작 원리와 개념을 누구보다 잘 알 수 있게 되겠죠?! 이렇게 얻은 이해를 바탕으로, 기존 발명품을 더욱 디벨롶하여 새로운 발명품을 만들어낼 기회도 생길 수 있을 것 같네요!! 이러한 점만 봤을 때는 재발명이 최고로 좋을 것 같지만,,,

단점도 존재해요! 우리는 원래 그 발명품을 만든 발명가가 아니기 때문에, 동일하게 재현하지 못할 가능성이 커요. 이는 시간적 비용이 많이 들 수 있으며, 그 노력에 대한 보상을 받지 못할 수도 있어요. 또한, 발명품의 동작 원리를 알 필요 없이 그저 잘 활용하는 것이 더 중요할 수 있어요!! 결국, 재발명의 필요성에 의문을 제기할 수밖에 없죠!! 예를 들어, 운전자는 자동차의 구동 원리를 알지 못하더라도, 안전하게 운전하는 것만으로 충분한 것처럼요!! (좋은 예시인지는 모르겠지만..)

음음 결론적으로, 재발명은 기존 발명품의 기능적 한계를 개선하거나 새로운 목적에 맞게 재설계할 필요가 있을 때, 또는 기술적 이해를 깊이 하고자 할 때 적합할 것 같아요! 그 외의 경우에는 발명품을 효과적으로 활용하는 데 집중하는 것이 더 나을 것 같아요!!

**Comment by Yunseok-Nam** [comment]
프로젝트를 진행하면서 새로운 라이브러리를 사용해 보았습니다. 해당 라이브러리의 내부가 어떻게 구현됐는지 알지 못해도 간단한 구현은 문제없이 할 수 있었습니다. 그러나 라이브러리의 내부 동작을 모르다 보니 기능을 확장하는 데 어려움을 느낀 경험이 있습니다. 그리고 내부 동작을 정확히 알지 못하니 왜 이 라이브러리를 사용해야 하는지에 대해 명확한 근거를 제시하기도 어려웠습니다.
이 경험을 바탕으로 제가 생각하는 바퀴를 재발명했을 때의 장점은 내부 동작을 이해함으로써 기능 확장이나 예상치 못한 상황에서의 대처에 유리할 것 같습니다. 또한 언제 해당 바퀴를 사용하는 것이 좋은지 자신만의 근거가 생길  것 같습니다.
단점으로는 바퀴를 재발명하기 위해 어디까지 갈지 모른다는 것입니다. 자칫 잘못하면 주어진 시간과 비용을 비효율적으로 보내게 될 것 같습니다.

**Comment by s6m1n** [comment]
## 재발명의 이점

### 학습
널리 사용되어 온 발명품은 여러 사람들의 시행착오 끝에 아마도 가장 최선의 구조를 갖게 되었을 것이다.
이는 오랜 시간을 거치며 경험이 축적된 집단 지식의 집약체로 볼 수 있다.
따라서 피상적인 이론 지식을 받아들이는 것보다 이미 잘 만들어진 발명품을 직접 재현해보는 과정에서 얻는 효과가 더 클 것이다.
기술과 방법이 채택된 이유를 파악하는 과정은 원리 기반 학습에 큰 도움이 되며
개념을 구체적인 문제 해결로 연결하는 실습을 통해 원리와 메커니즘을 더 깊이 이해할 수 있을 것이다.
더불어 설계나 패턴을 선택하는 과정에서 맞닥뜨리게 되는 다양한 결정 지점들을 통해 창의적 사고를 확장할 수 있다.

### 혁신
`결함이 발견되지 않았기`때문에 `본질적인 결함이 없는 것`이라고 착각하고 있을 수도 있다.
기존의 발명품이 만들어진 당시의 제약 조건 내에서 최선의 구조였더라도,
시간이 흐르고 기술이 발전하면서 숨겨진 비효율성이나 문제점이 새롭게 드러날 수 있다.
재발명 과정에서 의문을 제기하고 기존의 틀에서 벗어나 새로운 접근 방식으로 최적화를 시도하는 것은 혁신의 시작이 될 수 있다.

## 재발명의 단점

### 비용
재발명은 많은 시간, 비용, 노력을 필요로 한다.
이미 최적화된 솔루션이 존재함에도 불구하고 처음부터 새롭게 만들려면 개발 과정에서의 리소스 낭비가 발생할 수 있다.
또한 항상 재발명의 성공이 보장되지 않기에 효율성과 비용 측면에서 큰 손해를 입을 수 있다.
학습이 아닌 **실사용 목적의 재발명일 경우 이미 존재하는 기술을 활용했을 때보다 자원을 낭비할 확률이 더 크다.**

### 단순 복제
재발명은 무에서 완전히 새로운 유를 창조하는 것이 아니다.
기존에 나와 있는 원리와 방식을 참고하다 보면 자연스럽게 그 영향을 받게 되는데 이 과정에서 닫힌 사고가 발생할 가능성이 있다.
어떠한 의문 제기 없이 무의식적으로 기존 구조를 답습하게 되면 오히려 기존 문제점까지 반복하게 되는 상황이 생길 수 있다.
결국 **새로운 문제 해결로 이어지지 않고 단순 복제에 그쳐 실질적인 발전이 없는 결과가 도출될 위험이 있다.**



> 어떤 상황에서 바퀴를 재발명하는 행위가 필요할까요? 

실사용보단 학습을 목적으로 재발명을 시도해보자

**Comment by sh1mj1** [comment]
## 장점
### 학습
바퀴를 새로 만들면서 해당 기술의 근본적인 부분을 학습하기 좋다.
기술의 핵심을 직접 구현해가면서, 보다 더 깊은 이해를 할 수 있다.

### 훈련
유명한 프레임워크나 라이브러리를 직접 구현해가면서 sdk 개발의 훈련 연습을 할 수 있다.
sdk 개발도 서비스 개발과 크게 다를 것이 없다.
결국 OOP 등의 다형성이 중요하다.
그리고 익숙한 서비스 개발에서 벗어나 더 깊은 프로그래밍 방법에 대한 인사이트를 가질 수 있다. 

### 래퍼런스
이미 제공되어 있는 기능에서 조금 더 확장성이 필요한 sdk 를 개발해야 할 수 있다.
혹은, 더 많은 기능을 추가해야 할 수도 있다.
이러한 경우에 재발명, 리버스 엔지니어링을 통해 좋은 레퍼런스를 얻을 수 있다.

## 단점
### 시간과 자원의 낭비
이미 검증된 방법을 무시하고 처음부터 새롭게 재발명 하는 것은 예측하지 못한 오류나 발생 시킬 수 있다.
이를 해결하기 위해 많은 시간과 노력이 필요할 수 있다. 
기존의 방법이 이미 충분히 최적화된 상태일 것이다. 
새로운 해결책이 실제로 더 나은 결과를 가져온다는 보장이 없다.

코틀린은 확장 함수를 지원한다.
이를 통해 외부 라이브러리에도 기능을 쉽게 추가할 수 있다.
단순히 기능의 확장이 필요한 상황에서 재발명을 하는 것은 어리석다.
